discover.py

```diff
@@ -0,0 +1,234 @@
+
+SERVICE_TYPE = "_liquid._tcp.local."
+HOST_ONLY_NETWORK_MASK = '255.255.255.255'
+DNSMASQ_PROCESS_NAME = 'dnsmasq-dns'
+
+nodes = {}
+
+def reload_dnsmasq():
+    log.debug("Reloading dnsmasq-dns")
+    supervisor = supervisor_client.connect_unix_socket()
+    is_running = supervisor.getProcessInfo(DNSMASQ_PROCESS_NAME)['statename'] == 'RUNNING'
+    if is_running:
+        supervisor.stopProcess(DNSMASQ_PROCESS_NAME)
+    supervisor.startProcess(DNSMASQ_PROCESS_NAME)
+
+def stop_dnsmasq():
+    log.debug("Stopping dnsmasq-dns")
+    supervisor = supervisor_client.connect_unix_socket()
+    supervisor.stopProcess(DNSMSAQ_PROCESS_NAME)
+
+def rewrite_dnsmasq_conf(path, conf_string):
+    temp = path + ".tmp"
+    with open(temp, 'w') as f:
+        f.write(conf_string)
+    shutil.move(temp, path)
+
+class DnsmasqRestarter(Thread):
+    def __init__(self):
+        super().__init__()
+        self.last_dnsmasq_command = ''
+
+    def run(self):
+        while True:
+            self.update_dnsmasq()
+            time.sleep(3)
+
+    def update_dnsmasq(self):
+        interface = app.config['DNS_INTERFACE']
+        if not interface:
+            if self.last_dnsmasq_command:
+                stop_dnsmasq()
+                self.last_dnsmasq_command = ''
+            return
+
+        command = get_dns_conf_string(interface)
+        if command == self.last_dnsmasq_command:
+            return
+
+        rewrite_dnsmasq_conf(app.config['DNS_CONF_FILE'], command)
+        reload_dnsmasq()
+        self.last_dnsmasq_command = command
+
+
+def get_dns_conf_string(interface):
+    ip = get_ipv4_addr(interface)
+    common_args = [
+        'domain-needed',
+        'bogus-priv',
+        'server=208.67.222.222',
+        'server=208.67.220.220',
+        'no-resolv',
+        'no-hosts',
+        'bind-interfaces',
+    ]
+    listen_args = [
+        'interface={}'.format(interface),
+        'listen-address={}'.format(ip),
+    ]
+
+    address_records = [
+        (node['hostname'], node['address'])
+        for node in nodes.get(interface, {}).values()
+        if not node['is_local']
+    ]
+    address_records.sort()
+    address_records.insert(0, (app.config['LIQUID_DOMAIN'], ip))
+    address_args = [
+        'address=/{}/{}'.format(*record)
+        for record in address_records
+    ]
+
+    args = common_args + listen_args + address_args
+    return "\n".join(args) + "\n"
+
+def dict_decode(data):
+    def decode(data):
+        if isinstance(data, bytes):
+            return data.decode("latin-1")
+        elif isinstance(data, str):
+            return data
+        return str(data)
+
+    return {
+        decode(key): decode(data[key])
+        for key in data
+    }
+
+def get_ipv4_addr(interface):
+    addrs = netifaces.ifaddresses(interface).get(netifaces.AF_INET)
+    if not addrs:
+        return None
+
+    ips = [
+        a['addr']
+        for a in addrs
+        if 'addr' in a and a.get('netmask') != HOST_ONLY_NETWORK_MASK
```

Comment:
Here's an idea to make the code more readable. Take this bit, `'addr' in a and a.get('netmask') != HOST_ONLY_NETWORK_MASK`, and move it into a lambda. I'm not sure what that bit does. Let's assume it checks for unicorns. Then it can be `has_unicorns = lambda a: 'addr' in a and a.get('netmask') != HOST_ONLY_NETWORK_MASK`, and then the list comprehension becomes `ips = [a['addr'] for a in addrs if has_unicorns(a)]`.